watertap/unit_models/crystallizer.py

```diff
@@ -0,0 +1,696 @@
+        def eq_solubility_massfrac_equality_constraint(b, j):
+            if j in solute_set:
+                return (b.properties_out[0].mass_frac_phase_comp['Liq', j] - b.properties_out[0].solubility_mass_frac_phase_comp['Liq', j] == 0)
+            else:
+                return Constraint.Skip
+
+
+        # 3. Performance equations 
+        # (a) based on yield
+        @self.Constraint(self.config.property_package.component_list,
+                         doc="Component salt yield equation")
+        def eq_removal_balance(b, j):
+            if j in solvent_set:
+                return Constraint.Skip
+            else:
+                return (b.properties_in[0].flow_mass_phase_comp['Liq', j] * b.crystallization_yield[j]
+                    == b.properties_in[0].flow_mass_phase_comp['Liq', j] - b.properties_out[0].flow_mass_phase_comp['Liq', j])
+
+        # (b) Volumetric fraction constraint
+        @self.Constraint(doc="Solid volumetric fraction in outlet: constraint, 1-E")
+        def eq_vol_fraction_solids(b):
+            return (self.product_volumetric_solids_fraction 
+                == b.properties_solids[0].flow_vol / (b.properties_solids[0].flow_vol + b.properties_out[0].flow_vol)
+                )
+
+        # (c) Magma density constraint
+        @self.Constraint(doc="Slurry magma density")
+        def eq_dens_magma(b):
+            return (self.magma_density == 
+                b.properties_solids[0].dens_mass_solute['Sol'] * self.product_volumetric_solids_fraction
+                )
+
+        # (d) Operating pressure constraint
+        @self.Constraint(doc="Operating pressure constraint")
+        def eq_operating_pressure_constraint(b):
+            return (self.pressure_operating - b.properties_out[0].pressure_sat == 0)
+
+        # (e) Relative supersaturation 
+        @self.Constraint(solute_set, doc="Relative supersaturation created via evaporation, g/g (solution)")
+        def eq_relative_supersaturation(b, j):
+            #  mass_frac_after_evap = SOLIDS IN + LIQUID IN - VAPOUR OUT
+            mass_frac_after_evap = b.properties_in[0].flow_mass_phase_comp['Liq', j] \
+            / (sum(b.properties_in[0].flow_mass_phase_comp['Liq', j] for j in solute_set)
+                + b.properties_in[0].flow_mass_phase_comp['Liq', 'H2O']
+                - b.properties_vapor[0].flow_mass_phase_comp['Vap', 'H2O']
+                )
+            # return (b.relative_supersaturation[j] * b.properties_out[0].solubility_mass_frac_phase_comp['Liq', j] == 
+            # (mass_frac_after_evap - b.properties_out[0].solubility_mass_frac_phase_comp['Liq', j])
+            # )
+            return (b.relative_supersaturation[j] == 
+                (mass_frac_after_evap - b.properties_out[0].solubility_mass_frac_phase_comp['Liq', j])
+                / b.properties_out[0].solubility_mass_frac_phase_comp['Liq', j]
+                )
```

Comment:
I think there is an issue here. `j` is fed into the function, but this is changed in the sum with `for j in solute_set`.

Reply:
Good point. Changed just in case.